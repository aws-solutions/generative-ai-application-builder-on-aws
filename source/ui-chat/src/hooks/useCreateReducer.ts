/**********************************************************************************************************************
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.                                                *
 *                                                                                                                    *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance    *
 *  with the License. A copy of the License is located at                                                             *
 *                                                                                                                    *
 *      http://www.apache.org/licenses/LICENSE-2.0                                                                    *
 *                                                                                                                    *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES *
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions    *
 *  and limitations under the License.                                                                                *
 *********************************************************************************************************************/

import { useMemo, useReducer } from 'react';

// Extracts property names from initial state of reducer to allow typesafe dispatch objects
export type FieldNames<T> = {
    [K in keyof T]: T[K] extends string ? K : K;
}[keyof T];

// Returns the Action Type for the dispatch object to be used for typing in things like context
export type ActionType<T> = { type: 'reset' } | { type?: 'change'; field: FieldNames<T>; value: any };

// Returns a typed dispatch and state
export const useCreateReducer = <T>({ initialState }: { initialState: T }) => {
    type Action = { type: 'reset' } | { type?: 'change'; field: FieldNames<T>; value: any };

    const reducer = (state: T, action: Action) => {
        if (!action.type) return { ...state, [action.field]: action.value };

        if (action.type === 'reset') return initialState;

        throw new Error();
    };

    const [state, dispatch] = useReducer(reducer, initialState);

    return useMemo(() => ({ state, dispatch }), [state, dispatch]);
};
