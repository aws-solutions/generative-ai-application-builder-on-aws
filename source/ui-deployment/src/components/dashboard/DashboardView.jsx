/**********************************************************************************************************************
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.                                                *
 *                                                                                                                    *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance    *
 *  with the License. A copy of the License is located at                                                             *
 *                                                                                                                    *
 *      http://www.apache.org/licenses/LICENSE-2.0                                                                    *
 *                                                                                                                    *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES *
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions    *
 *  and limitations under the License.                                                                                *
 *********************************************************************************************************************/

import { useRef, useState, useEffect, useContext } from 'react';
import { useNavigate } from 'react-router-dom';
import { Button, Link, Pagination, StatusIndicator, Table, TextFilter } from '@cloudscape-design/components';
import { useCollection } from '@cloudscape-design/collection-hooks';
import { FullPageHeader } from '../commons';
import { Breadcrumbs, ToolsContent } from './common-components';
import {
    CustomAppLayout,
    Navigation,
    Notifications,
    TableEmptyState,
    TableNoMatchState
} from '../commons/common-components';
import {
    paginationAriaLabels,
    deploymentTableAriaLabels,
    renderAriaLive,
    getTextFilterCounterText,
    getHeaderCounterText,
    createTableSortLabelFn
} from '../../i18n-strings';
import { useColumnWidths } from '../commons/use-column-widths';
import { useLocalStorage } from '../commons/use-local-storage';
import HomeContext from '../../home/home.context';
import { DEFAULT_PREFERENCES, Preferences, parseStackName } from '../commons/table-config';
import { listDeployedUseCases, statusIndicatorTypeSelector } from './deployments';
import { DeleteDeploymentModal, onDeleteConfirm } from '../commons/delete-modal';
import { dateOptions } from '../useCaseDetails/common-components';
import { CFN_STACK_STATUS_INDICATOR, ERROR_MESSAGES } from '../../utils/constants';

function UseCaseTable({ columnDefinitions, saveWidths, loadHelpPanelContent }) {
    const [selectedItems, setSelectedItems] = useState([]);
    const [, setFilteringText] = useState('');
    const [, setDelayedFilteringText] = useState('');
    const [deployments, setDeployments] = useState([]);
    const [loading, setLoading] = useState(false);
    const [showDeleteModal, setShowDeleteModal] = useState(false);
    const [preferences, setPreferences] = useLocalStorage('Preferences', DEFAULT_PREFERENCES);

    const onDeleteInit = () => setShowDeleteModal(true);
    const onDeleteDiscard = () => setShowDeleteModal(false);

    const {
        state: { deploymentsData, reloadData },
        dispatch: homeDispatch
    } = useContext(HomeContext);

    const fetchData = async (isReload = false) => {
        try {
            setLoading(true);

            if (deploymentsData.length === 0 || reloadData || isReload) {
                const response = await listDeployedUseCases();
                setDeployments(response.deployments);
                homeDispatch({
                    field: 'deploymentsData',
                    value: response.deployments
                });
            } else {
                setDeployments(deploymentsData);
            }
            setLoading(false);
            homeDispatch({
                field: 'reloadData',
                value: false
            });
            homeDispatch({
                field: 'authorized',
                value: true
            });
        } catch (error) {
            if (error.message === ERROR_MESSAGES.UNAUTHORIZED) {
                homeDispatch({
                    field: 'authorized',
                    value: false
                });
            }
            console.error(error);
        }
    };

    useEffect(() => {
        fetchData();
    }, [reloadData]);

    const dateCreatedColumn = columnDefinitions.find((column) => column.id === 'dateCreated');

    const { items, actions, filteredItemsCount, collectionProps, filterProps, paginationProps } = useCollection(
        deployments,
        {
            filtering: {
                empty: <TableEmptyState resourceName="Deployment" />,
                noMatch: <TableNoMatchState onClearFilter={() => actions.setFiltering('')} />
            },
            pagination: { pageSize: preferences.pageSize },
            sorting: { defaultState: { sortingColumn: dateCreatedColumn, isDescending: true } },
            selection: {}
        }
    );

    const onClearFilter = () => {
        setFilteringText('');
        setDelayedFilteringText('');
    };

    const onSelectionChange = ({ detail }) => {
        setSelectedItems(detail.selectedItems);
        homeDispatch({
            field: 'selectedDeployment',
            value: detail.selectedItems[0]
        });
    };

    return (
        <div>
            <Table
                {...collectionProps}
                loading={loading}
                items={items}
                onSelectionChange={onSelectionChange}
                selectedItems={selectedItems}
                columnDefinitions={columnDefinitions}
                columnDisplay={preferences.contentDisplay}
                selectionType="single"
                ariaLabels={deploymentTableAriaLabels}
                renderAriaLive={renderAriaLive}
                variant="full-page"
                stickyHeader={true}
                resizableColumns={true}
                onColumnWidthsChange={saveWidths}
                wrapLines={preferences.wrapLines}
                stripedRows={preferences.stripedRows}
                contentDensity={preferences.contentDensity}
                stickyColumns={preferences.stickyColumns}
                header={
                    <FullPageHeader
                        counter={!loading && getHeaderCounterText(deployments, collectionProps.selectedItems)}
                        onInfoLinkClick={loadHelpPanelContent}
                        selectedItems={selectedItems}
                        setSelectedItems={setSelectedItems}
                        refreshData={fetchData}
                        onDeleteInit={onDeleteInit}
                    />
                }
                loadingText="Loading deployments"
                empty={<TableNoMatchState onClearFilter={onClearFilter} />}
                filter={
                    <TextFilter
                        {...filterProps}
                        filteringAriaLabel="Filter deployments"
                        filteringPlaceholder="Find deployments"
                        filteringClearAriaLabel="Clear"
                        countText={getTextFilterCounterText(filteredItemsCount)}
                    />
                }
                pagination={
                    <Pagination {...paginationProps} ariaLabels={paginationAriaLabels(filteredItemsCount.pagesCount)} />
                }
                preferences={<Preferences preferences={preferences} setPreferences={setPreferences} />}
            />

            <DeleteDeploymentModal
                visible={showDeleteModal}
                onDiscard={onDeleteDiscard}
                onDelete={onDeleteConfirm}
                deployment={selectedItems[0]}
            />
        </div>
    );
}

const createCloudfrontUrlLinkComponent = (item) => {
    if (!item.cloudFrontWebUrl || statusIndicatorTypeSelector(item.status) !== CFN_STACK_STATUS_INDICATOR.SUCCESS) {
        return <div>{'-'}</div>;
    }

    return (
        <div>
            <Link external href={`${item.cloudFrontWebUrl}`}>
                {'Open Application'}
            </Link>
        </div>
    );
};

export default function DashboardView() {
    const { dispatch: homeDispatch } = useContext(HomeContext);
    const navigate = useNavigate();
    const [toolsOpen, setToolsOpen] = useState(false);

    const handleOnDeploymentIdClick = (deploymentItem) => {
        homeDispatch({
            field: 'selectedDeployment',
            value: deploymentItem
        });
        navigate(`/deployment-details`);
    };

    const createDetailsPageLink = (item) => {
        return (
            <Button variant="link" onClick={() => handleOnDeploymentIdClick(item)}>
                {parseStackName(item.Name)}
            </Button>
        );
    };

    const displayStackStatus = (item) => {
        const cleanStatusString = (status) => {
            return status.replaceAll('_', ' ').toLowerCase();
        };
        return (
            <StatusIndicator type={statusIndicatorTypeSelector(item.status)}>
                {cleanStatusString(item.status)}
            </StatusIndicator>
        );
    };

    const rawColumns = [
        {
            id: 'name',
            sortingField: 'Name',
            cell: (item) => createDetailsPageLink(item),
            header: 'Use Case Name',
            minWidth: 160,
            isRowHeader: true
        },
        {
            id: 'stackId',
            sortingField: 'useCaseUUID',
            header: 'Deployment Stack ID',
            cell: (item) => item.useCaseUUID,
            minWidth: 100
        },
        {
            id: 'status',
            sortingField: 'status',
            header: 'Application Status',
            cell: (item) => displayStackStatus(item),
            minWidth: 60
        },
        {
            id: 'dateCreated',
            sortingField: 'CreatedDate',
            header: 'Date Created',
            cell: (item) => new Date(item.CreatedDate).toLocaleDateString('en-US', dateOptions),
            minWidth: 120
        },
        {
            id: 'modelProvider',
            sortingField: 'ModelProvider',
            header: 'Model Provider',
            cell: (item) => item.LlmParams.ModelProvider ?? 'N/A',
            minWidth: 100
        },
        {
            id: 'webUrl',
            sortingField: 'cloudFrontWebUrl',
            header: 'Application Access',
            cell: (item) => createCloudfrontUrlLinkComponent(item),
            minWidth: 120
        }
    ];

    const COLUMN_DEFINITIONS = rawColumns.map((column) => ({
        ...column,
        ariaLabel: createTableSortLabelFn(column)
    }));

    const [columnDefinitions, saveWidths] = useColumnWidths('DeploymentsDashboard-Widths', COLUMN_DEFINITIONS);

    const appLayout = useRef();

    const onFollowNavigationHandler = (event) => {
        navigate(event.detail.href);
    };

    return (
        <div>
            <CustomAppLayout
                ref={appLayout}
                navigation={<Navigation onFollowHandler={onFollowNavigationHandler} />}
                breadcrumbs={<Breadcrumbs />}
                content={
                    <UseCaseTable
                        columnDefinitions={columnDefinitions}
                        saveWidths={saveWidths}
                        loadHelpPanelContent={() => {
                            setToolsOpen(true);
                            appLayout.current?.focusToolsClose();
                        }}
                        setSelectedDeploymentStorage
                    />
                }
                contentType="table"
                tools={<ToolsContent />}
                toolsOpen={toolsOpen}
                onToolsChange={({ detail }) => setToolsOpen(detail.open)}
                notifications={<Notifications successNotification={true} />}
                stickyNotifications
                data-testid="dashboard-view"
            />
        </div>
    );
}
