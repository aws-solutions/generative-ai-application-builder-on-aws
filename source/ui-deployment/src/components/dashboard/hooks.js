/**********************************************************************************************************************
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.                                                *
 *                                                                                                                    *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance    *
 *  with the License. A copy of the License is located at                                                             *
 *                                                                                                                    *
 *      http://www.apache.org/licenses/LICENSE-2.0                                                                    *
 *                                                                                                                    *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES *
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions    *
 *  and limitations under the License.                                                                                *
 *********************************************************************************************************************/

import { useEffect, useState } from 'react';
import { fetchDeployments } from './deployments';

export function useDeployments(params = {}) {
    const { pageSize, currentPageIndex: clientPageIndex } = params.pagination || {};
    const { sortingDescending, sortingColumn } = params.sorting || {};
    const { filteringText, filteringTokens, filteringOperation } = params.filtering || {};
    const [loading, setLoading] = useState(false);
    const [items, setItems] = useState([]);
    const [totalCount, setTotalCount] = useState(0);
    const [currentPageIndex, setCurrentPageIndex] = useState(clientPageIndex);
    const [pagesCount, setPagesCount] = useState(0);

    useEffect(() => {
        setCurrentPageIndex(clientPageIndex);
    }, [clientPageIndex]);

    useEffect(() => {
        setLoading(true);
        const params = {
            filteringText,
            filteringTokens,
            filteringOperation,
            pageSize,
            currentPageIndex,
            sortingDescending,
            ...(sortingColumn
                ? {
                      sortingColumn: sortingColumn.sortingField
                  }
                : {})
        };
        const callback = ({ items, pagesCount, currentPageIndex }) => {
            setLoading(false);
            setItems(items);
            setPagesCount(pagesCount + 1);
            setCurrentPageIndex(currentPageIndex);
            if (totalCount === 0) {
                setTotalCount(pagesCount > 1 ? pageSize * (pagesCount - 1) : items.length);
            }
        };
        fetchDeployments(params, callback);
    }, [
        pageSize,
        sortingDescending,
        sortingColumn,
        currentPageIndex,
        filteringText,
        filteringTokens,
        filteringOperation
    ]);

    return {
        items,
        loading,
        totalCount,
        pagesCount,
        currentPageIndex
    };
}
