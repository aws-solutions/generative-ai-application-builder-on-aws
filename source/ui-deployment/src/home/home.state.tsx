/**********************************************************************************************************************
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.                                                *
 *                                                                                                                    *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance    *
 *  with the License. A copy of the License is located at                                                             *
 *                                                                                                                    *
 *      http://www.apache.org/licenses/LICENSE-2.0                                                                    *
 *                                                                                                                    *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES *
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions    *
 *  and limitations under the License.                                                                                *
 *********************************************************************************************************************/

import { DEPLOYMENT_ACTIONS } from '../utils/constants';

export interface HomeInitialState {
    authorized: boolean;
    selectedDeployment: any;
    deploymentsData: any;
    deploymentAction: string;
    runtimeConfig?: any;
    reloadData?: boolean;
}

export const initialState: HomeInitialState = {
    authorized: true,
    selectedDeployment: {},
    deploymentsData: [],
    deploymentAction: DEPLOYMENT_ACTIONS.CREATE,
    reloadData: false
};

export const insertRuntimeConfig = (state: Partial<HomeInitialState>, runtimeConfig: any) => {
    return {
        ...state,
        runtimeConfig
    };
};
