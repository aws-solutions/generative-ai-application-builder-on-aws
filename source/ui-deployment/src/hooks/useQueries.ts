// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import { MODEL_PROVIDER_NAME_MAP } from '@/components/wizard/steps-config';
import { BEDROCK_INFERENCE_TYPES } from '@/utils/constants';
import { useQuery } from '@tanstack/react-query';
import {
    fetchModelDefaults,
    fetchModelIds,
    fetchModelProviders,
    fetchModelStreamingDefault,
    fetchModelTemperatureRange,
    setDefaultQueryParams,
    fetchUseCaseDetails
} from 'services';

const INFERENCE_PROFILE = 'inference-profile';

/**
 * Helper function to determine the appropriate modelId to use for API calls
 * based on the provider and inference type
 * @param providerName LLM provider name
 * @param modelId Original model ID
 * @param bedrockInferenceType Bedrock inference type if applicable
 * @returns The model ID to use for API calls
 */
export const getEffectiveModelId = (providerName: string, modelId: string, bedrockInferenceType?: string): string => {
    // Only apply special handling for Bedrock provider
    if (providerName === MODEL_PROVIDER_NAME_MAP.Bedrock && bedrockInferenceType) {
        // For non-QuickStart inference types, use the inference-profile
        if (
            bedrockInferenceType === BEDROCK_INFERENCE_TYPES.OTHER_FOUNDATION_MODELS ||
            bedrockInferenceType === BEDROCK_INFERENCE_TYPES.INFERENCE_PROFILES ||
            bedrockInferenceType === BEDROCK_INFERENCE_TYPES.PROVISIONED_MODELS
        ) {
            return INFERENCE_PROFILE;
        }
    }

    // For all other cases, use the provided modelId
    return modelId;
};

/**
 * Returns a hook that fetches the available model names for provider, and caches the
 * responses
 * @param providerName LLM provider name
 * @param useCaseType Defaults to 'Chat'
 * @returns
 */
export const useModelNameQuery = (providerName: string, useCaseType?: string) => {
    const fetchParams = setDefaultQueryParams({ providerName, useCaseType });
    return useQuery({
        queryKey: ['modelNames', fetchParams.providerName, fetchParams.useCaseType],
        queryFn: async () => await fetchModelIds(fetchParams),
        retry: 10
    });
};

/**
 * Returns a hook that fetches the model defaults for provider, and caches the response
 * @param providerName LLM provider name
 * @param modelId Model ID to fetch defaults for. Defaults to first model in list.
 * @param useCaseType Defaults to 'Chat'
 * @param bedrockInferenceType Bedrock inference type if applicable
 * @returns
 */
export const useModelInfoQuery = (
    providerName: string,
    modelId: string,
    bedrockInferenceType: string,
    useCaseType?: string
) => {
    // Get the effective model ID for the API call
    const effectiveModelId = getEffectiveModelId(providerName, modelId, bedrockInferenceType);

    // Determine when the query should be enabled
    // 1. For SageMaker, always enable the query
    // 2. For Bedrock with special inference types, enable even with empty modelId
    // 3. For other cases, require both providerName and modelId
    const isSpecialBedrockInferenceType = 
        providerName === MODEL_PROVIDER_NAME_MAP.Bedrock && 
        (bedrockInferenceType === BEDROCK_INFERENCE_TYPES.INFERENCE_PROFILES || 
         bedrockInferenceType === BEDROCK_INFERENCE_TYPES.PROVISIONED_MODELS ||
         bedrockInferenceType === BEDROCK_INFERENCE_TYPES.OTHER_FOUNDATION_MODELS);
    
    const enabled = 
        providerName === MODEL_PROVIDER_NAME_MAP.SageMaker || 
        isSpecialBedrockInferenceType || 
        Boolean(providerName && modelId);
    
    const fetchParams = setDefaultQueryParams({ modelId: effectiveModelId, providerName, useCaseType });
    return useQuery({
        queryKey: [
            'modelInfo',
            fetchParams.providerName,
            fetchParams.useCaseType,
            effectiveModelId,
            bedrockInferenceType
        ],
        queryFn: async () => await fetchModelDefaults(fetchParams),
        retry: 10,
        enabled: enabled // Only run the query when both provider and model ID are available
    });
};

/**
 * Query to retrieve model streaming info
 * @param providerName LLM provider name
 * @param modelId Model ID to fetch defaults for. Defaults to first model in list.
 * @param useCaseType Defaults to 'Chat'
 * @param bedrockInferenceType Bedrock inference type if applicable
 * @returns
 */
export const useModelStreamingQuery = (
    providerName: string,
    modelId: string,
    bedrockInferenceType: string,
    useCaseType?: string
) => {
    // Get the effective model ID for the API call
    const effectiveModelId = getEffectiveModelId(providerName, modelId, bedrockInferenceType);

    const fetchParams = setDefaultQueryParams({ modelId: effectiveModelId, providerName, useCaseType });
    return useQuery({
        queryKey: ['streaming', fetchParams.providerName, effectiveModelId, bedrockInferenceType],
        queryFn: async () => await fetchModelStreamingDefault(fetchParams),
        retry: 10
    });
};

/**
 * Query hook to get default model temperature
 * @param providerName LLM provider name
 * @param modelId Model ID to fetch defaults for. Defaults to first model in list.
 * @param useCaseType Defaults to 'Chat'
 * @param bedrockInferenceType Bedrock inference type if applicable
 * @returns
 */
export const useModelTemperatureQuery = (
    providerName: string,
    modelId: string,
    bedrockInferenceType: string,
    useCaseType?: string
) => {
    // Get the effective model ID for the API call
    const effectiveModelId = getEffectiveModelId(providerName, modelId, bedrockInferenceType);

    const fetchParams = setDefaultQueryParams({ modelId: effectiveModelId, providerName, useCaseType });
    return useQuery({
        queryKey: ['temperature', fetchParams.providerName, effectiveModelId, bedrockInferenceType],
        queryFn: async () => await fetchModelTemperatureRange(fetchParams),
        retry: 10
    });
};

/**
 * Get the supported model providers
 * @param useCaseType Defaults to 'Chat'
 * @returns
 */
export const useModelProvidersQuery = (useCaseType?: string) => {
    const fetchParams = setDefaultQueryParams({ useCaseType });

    return useQuery({
        queryKey: ['modelProviders', fetchParams.useCaseType],
        queryFn: async () => await fetchModelProviders(fetchParams),
        retry: 10
    });
};

/**
 * Returns a hook that fetches the details of a specific use case and caches the response
 * @param useCaseId The ID of the use case to fetch
 * @param options Optional configuration options for the query
 * @returns A query result object containing the use case details
 */
export const useUseCaseDetailsQuery = (
    useCaseId: string,
    options?: {
        enabled?: boolean;
        refetchInterval?: number | false;
        refetchOnWindowFocus?: boolean;
    }
) => {
    return useQuery({
        queryKey: ['useCaseDetails', useCaseId],
        queryFn: async () => {
            if (!useCaseId) {
                throw new Error('Missing useCaseId');
            }
            return await fetchUseCaseDetails({ useCaseId });
        },
        enabled: options?.enabled !== false && !!useCaseId,
        refetchInterval: options?.refetchInterval,
        refetchOnWindowFocus: options?.refetchOnWindowFocus,
        retry: 3,
        staleTime: 5 * 60 * 1000 // 5 minutes
    });
};
