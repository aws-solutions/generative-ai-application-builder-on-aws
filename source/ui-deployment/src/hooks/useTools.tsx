// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import { HelpPanel, AppLayoutProps } from '@cloudscape-design/components';
import { useState, useRef } from 'react';
import { ExternalLinkGroup } from '../components/commons';
import { ToolHelpPanelContent } from '../components/wizard/interfaces/Steps';
import { BaseWizardStep } from '@/components/wizard/interfaces/Steps/BaseWizardStep';

const getFormattedToolsContent = (tools: ToolHelpPanelContent) => (
    <HelpPanel header={<h2>{tools.title}</h2>} footer={tools.links && <ExternalLinkGroup items={tools.links} data-testid="help-panel"
    />}>
        {tools.content}
    </HelpPanel>
);

const generateTools = (content: ToolHelpPanelContent) => {
    const [toolsContent, setToolsContent] = useState(getFormattedToolsContent(content));
    const [isToolsOpen, setIsToolsOpen] = useState(false);
    const appLayoutRef = useRef<AppLayoutProps.Ref>(null);

    const setFormattedToolsContent = (tools: ToolHelpPanelContent) => {
        setToolsContent(getFormattedToolsContent(tools));
    };

    const setHelpPanelContent = (tools: ToolHelpPanelContent) => {
        if (tools) {
            setFormattedToolsContent(tools);
        }
        setIsToolsOpen(true);
        appLayoutRef.current?.focusToolsClose();
    };
    const closeTools = () => setIsToolsOpen(false);

    const onToolsChange = (evt: AppLayoutProps.ChangeDetail) => setIsToolsOpen(evt.open);

    return {
        content: toolsContent,
        isOpen: isToolsOpen,
        setContentAndOpen: setHelpPanelContent,
        close: closeTools,
        setContent: setFormattedToolsContent,
        onChange: onToolsChange,
        appLayoutRef: appLayoutRef
    };
};

export const generateToolsForStep = (step: BaseWizardStep) => {
    return generateTools(step.toolContent);
};
