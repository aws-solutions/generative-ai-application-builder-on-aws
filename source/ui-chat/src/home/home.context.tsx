/**********************************************************************************************************************
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.                                                *
 *                                                                                                                    *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance    *
 *  with the License. A copy of the License is located at                                                             *
 *                                                                                                                    *
 *      http://www.apache.org/licenses/LICENSE-2.0                                                                    *
 *                                                                                                                    *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES *
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions    *
 *  and limitations under the License.                                                                                *
 *********************************************************************************************************************/

import { Dispatch, createContext } from 'react';

import { ActionType } from '../hooks/useCreateReducer';

import { Conversation } from '../types/chat';
import { KeyValuePair } from '../types/misc';

import { HomeInitialState } from './home.state';

export interface HomeContextProps {
    state: HomeInitialState;
    dispatch: Dispatch<ActionType<HomeInitialState>>;
    handleUpdateConversation: (conversation: Conversation, data: KeyValuePair) => void;
}

const HomeContext = createContext<HomeContextProps>(undefined!);

export default HomeContext;
