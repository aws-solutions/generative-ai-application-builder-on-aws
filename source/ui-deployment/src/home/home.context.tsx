/**********************************************************************************************************************
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.                                                *
 *                                                                                                                    *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance    *
 *  with the License. A copy of the License is located at                                                             *
 *                                                                                                                    *
 *      http://www.apache.org/licenses/LICENSE-2.0                                                                    *
 *                                                                                                                    *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES *
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions    *
 *  and limitations under the License.                                                                                *
 *********************************************************************************************************************/

import { Dispatch, createContext, useEffect } from 'react';

import { ActionType } from '../hooks/useCreateReducer';

import { HomeInitialState } from './home.state';

export interface HomeContextProps {
    state: HomeInitialState;
    dispatch: Dispatch<ActionType<HomeInitialState>>;
}

const HomeContext = createContext<HomeContextProps>(undefined!);

export const HomeContextProvider = (props: any) => {
    useEffect(() => {
        sessionStorage.setItem('init-state', JSON.stringify(props.value.state));
    }, [props.value.state]);

    return <HomeContext.Provider value={props.value}>{props.children}</HomeContext.Provider>;
};

export default HomeContext;
