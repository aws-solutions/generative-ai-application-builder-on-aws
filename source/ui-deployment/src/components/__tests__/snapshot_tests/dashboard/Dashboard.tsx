/**********************************************************************************************************************
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.                                                *
 *                                                                                                                    *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance    *
 *  with the License. A copy of the License is located at                                                             *
 *                                                                                                                    *
 *      http://www.apache.org/licenses/LICENSE-2.0                                                                    *
 *                                                                                                                    *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES *
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions    *
 *  and limitations under the License.                                                                                *
 *********************************************************************************************************************/

import '@testing-library/jest-dom';
import { Dispatch } from 'react';
import { MemoryRouter, Route, Routes } from 'react-router-dom';
import renderer from 'react-test-renderer';
import HomeContext from '../../../../home/home.context';
import { ActionType } from '../../../../hooks/useCreateReducer';
import { HomeInitialState } from '../../../../home/home.state';
import DashboardView from '../../../dashboard/DashboardView';

// eslint-disable-next-line jest/no-mocks-import
import mockContext from '../../__mocks__/mock-context.json';

jest.mock('@cloudscape-design/components', () => {
    const Components = jest.genMockFromModule('@cloudscape-design/components') as any;
    for (const componentName of Object.keys(Components)) {
        Components[componentName] = componentName;
    }
    return Components;
});

const contextValue = {
    dispatch: jest.fn() as Dispatch<ActionType<HomeInitialState>>,
    state: mockContext
};

test('Snapshot test for Dashboard', async () => {
    const tree = renderer
        .create(
            <>
                <HomeContext.Provider
                    value={{
                        ...contextValue
                    }}
                >
                    <MemoryRouter initialEntries={['//']}>
                        <Routes>
                            <Route path="/" element={<DashboardView />} />
                        </Routes>
                    </MemoryRouter>
                </HomeContext.Provider>
            </>
        )
        .toJSON();
    expect(tree).toMatchSnapshot();
});
