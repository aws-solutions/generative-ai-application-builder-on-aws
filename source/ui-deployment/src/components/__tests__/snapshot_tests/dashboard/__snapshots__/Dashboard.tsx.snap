// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot test for Dashboard 1`] = `
<div>
  <AppLayout
    ariaLabels={
      Object {
        "navigation": "Side navigation",
        "navigationClose": "Close side navigation",
        "navigationToggle": "Open side navigation",
        "notifications": "Notifications",
        "tools": "Help panel",
        "toolsClose": "Close help panel",
        "toolsToggle": "Open help panel",
      }
    }
    breadcrumbs={<Breadcrumbs />}
    content={
      <UseCaseTable
        columnDefinitions={
          Array [
            Object {
              "ariaLabel": [Function],
              "cell": [Function],
              "header": "Use Case Name",
              "id": "name",
              "isRowHeader": true,
              "minWidth": 160,
              "sortingField": "Name",
              "width": undefined,
            },
            Object {
              "ariaLabel": [Function],
              "cell": [Function],
              "header": "Deployment Stack ID",
              "id": "stackId",
              "minWidth": 100,
              "sortingField": "useCaseUUID",
              "width": undefined,
            },
            Object {
              "ariaLabel": [Function],
              "cell": [Function],
              "header": "Application Status",
              "id": "status",
              "minWidth": 60,
              "sortingField": "status",
              "width": undefined,
            },
            Object {
              "ariaLabel": [Function],
              "cell": [Function],
              "header": "Date Created",
              "id": "dateCreated",
              "minWidth": 120,
              "sortingField": "CreatedDate",
              "width": undefined,
            },
            Object {
              "ariaLabel": [Function],
              "cell": [Function],
              "header": "Model Provider",
              "id": "modelProvider",
              "minWidth": 100,
              "sortingField": "ModelProvider",
              "width": undefined,
            },
            Object {
              "ariaLabel": [Function],
              "cell": [Function],
              "header": "Application Access",
              "id": "webUrl",
              "minWidth": 120,
              "sortingField": "cloudFrontWebUrl",
              "width": undefined,
            },
          ]
        }
        loadHelpPanelContent={[Function]}
        saveWidths={[Function]}
        setSelectedDeploymentStorage={true}
      />
    }
    contentType="table"
    data-testid="dashboard-view"
    navigation={
      <Navigation
        onFollowHandler={[Function]}
      />
    }
    notifications={
      <Notifications
        successNotification={true}
      />
    }
    onToolsChange={[Function]}
    stickyNotifications={true}
    tools={<ToolsContent />}
    toolsOpen={false}
  />
</div>
`;
